components/backup-stream/src/subscription_manager.rs

```diff
@@ -630,42 +659,81 @@ where
                 )
             })?;
         let new_region_info = rx
-            .recv()
+            .await
             .map_err(|err| annotate!(err, "BUG?: unexpected channel message dropped."))?;
         if new_region_info.is_none() {
             metrics::SKIP_RETRY
                 .with_label_values(&["region-absent"])
                 .inc();
-            return Ok(());
+            return Ok(false);
         }
         let new_region_info = new_region_info.unwrap();
         if new_region_info.role != StateRole::Leader {
             metrics::SKIP_RETRY.with_label_values(&["not-leader"]).inc();
-            return Ok(());
+            return Ok(false);
+        }
+        if raftstore::store::util::compare_region_epoch(
+            region.get_region_epoch(),
+            &new_region_info.region,
+            true,
+            true,
+            false,
+        )
+        .is_err()
+        {
+            metrics::SKIP_RETRY
+                .with_label_values(&["epoch-not-match"])
+                .inc();
+            return Ok(false);
         }
         // Note: we may fail before we insert the region info to the subscription map.
         // At that time, the command isn't steal and we should retry it.
         let mut exists = false;
-        let removed = self.subs.deregister_region_if(&region, |old, _| {
+        let removed = self.subs.deregister_region_if(region, |old, _| {
             exists = true;
             let should_remove = old.handle().id == handle.id;
             if !should_remove {
-                warn!("stale retry command"; utils::slog_region(&region), "handle" => ?handle, "old_handle" => ?old.handle());
+                warn!("stale retry command"; utils::slog_region(region), "handle" => ?handle, "old_handle" => ?old.handle());
             }
             should_remove
         });
         if !removed && exists {
             metrics::SKIP_RETRY
                 .with_label_values(&["stale-command"])
                 .inc();
-            return Ok(());
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    async fn retry_observe(&self, region: Region, handle: ObserveHandle) -> Result<bool> {
+        let failure_count = *self.failure_count.get(&region.id).unwrap_or_else(|| {
```

Comment:
perhaps move `*self.failure_count.entry(region_id).or_insert(0) += 1;` here. to avoid panic.